### ISSUE TITLE
BulletPhysics cannot be compiled with clang-cl toolset on Windows

### ISSUE THREAD

**Issue by RIscRIpt** [bulletphysics/bullet3#4007]
TL;DR: bugs are in clang-cl, I've created this issue so you would be aware of the problems and could observe upstream bugs:

* https://bugs.llvm.org/show_bug.cgi?id=32603
* https://bugs.llvm.org/show_bug.cgi?id=44641

---

Trying to compile BulletPhysics on Windows using clang-cl toolset results in "duplicate symbol errors" from lld-link; this bug is known upstream: https://bugs.llvm.org/show_bug.cgi?id=32603

```pwsh
> cmake -B build -T clangcl -A Win32
...
lld-link : error : duplicate symbol: public: __thiscall SimpleCamera::SimpleCamera(void)
lld-link : error : duplicate symbol: public: virtual __thiscall SimpleCamera::~SimpleCamera(void)
lld-link : error : duplicate symbol: public: virtual void __thiscall SimpleCamera::setVRCamera(float const *const, float const *const)
...
```

According to that bug report you can add linker flag `/force:multiple`  to mitigate the problem:

```pwsh
> $Env:LDFLAGS="/force:multiple"; cmake -B build -T clangcl -A Win32
...
lld-link : warning : duplicate symbol: public: __thiscall SimpleCamera::SimpleCamera(void)
lld-link : warning : duplicate symbol: public: virtual __thiscall SimpleCamera::~SimpleCamera(void)
lld-link : warning : duplicate symbol: public: virtual void __thiscall SimpleCamera::setVRCamera(float const *const, float const *const)
...
```

Using this workaround, compilation finishes successfully (and `ctest` even passes, but only if it is compiled as x64 binary), however, running `ctest` from build directory for x86 (Win32), results in several failures:
```
D:\tmp\bullet3-3.17\build> ctest
Test project D:/tmp/bullet3-3.17/build
    Start 1: Test_BulletInverseDynamicsJacobian_PASS
1/7 Test #1: Test_BulletInverseDynamicsJacobian_PASS ....   Passed    0.24 sec
    Start 2: Test_BulletInverseForwardDynamics_PASS
2/7 Test #2: Test_BulletInverseForwardDynamics_PASS .....***Failed    0.03 sec
    Start 3: Test_BulletInverseDynamics_PASS
3/7 Test #3: Test_BulletInverseDynamics_PASS ............   Passed    0.03 sec
    Start 4: Test_PhysicsClientServer_PASS
4/7 Test #4: Test_PhysicsClientServer_PASS ..............***Failed    0.05 sec
    Start 5: Test_Collision_PASS
5/7 Test #5: Test_Collision_PASS ........................***Failed    0.04 sec
    Start 6: Test_btKinematicCharacterController_PASS
6/7 Test #6: Test_btKinematicCharacterController_PASS ...   Passed    0.02 sec
    Start 7: Test_BulletDynamics_PASS
7/7 Test #7: Test_BulletDynamics_PASS ...................   Passed    0.03 sec
```

Inspecting all the binaries with failures:
* .\test\InverseDynamics\Release\Test_BulletInverseForwardDynamics.exe
* .\test\SharedMemory\Release\Test_PhysicsClientServer.exe
* .\test\collision\Release\Test_Collision.exe

Reveals the problem: clang-cl generates code which assumes aligned load/store of memory, which in fact is not aligned; an example for `Test_BulletInverseForwardDynamics.exe`:
```
001BDF74 | 0F2847 30           | movaps xmm0,xmmword ptr ds:[edi+30]; copy rhs.m_rgbaColor into xmm0
001BDF78 | 0F284F 40           | movaps xmm1,xmmword ptr ds:[edi+40]; copy rhs.m_specularColor into xmm1
001BDF7C | 0F2941 30           | movaps xmmword ptr ds:[ecx+30],xmm0; save xmm0 into lhs.m_rgbaColor <- crash is here
001BDF80 | 0F2949 40           | movaps xmmword ptr ds:[ecx+40],xmm1; save xmm1 into lhs.m_specularColor
```
where `ecx = 009F5EF8` (not aligned).

Relevant reported bug upstream: https://bugs.llvm.org/show_bug.cgi?id=44641

Inspecting Debug build of `Test_BulletInverseForwardDynamics.exe` shows that the problem is in copy constructor of `btVector3`:
* [UrdfParser.cpp:742](https://github.com/bulletphysics/bullet3/blob/296cd85/examples/Importers/ImportURDFDemo/UrdfParser.cpp#L742)
* (implicitly defined copy constructor of `UrdfMaterialColor`)
* Inlined copy constructor of `btVector3` [btVector3.h:142](https://github.com/bulletphysics/bullet3/blob/296cd85/src/LinearMath/btVector3.h#L142)

Part of implicitly defined copy constructor of `UrdfMaterialColor` is listed above.

**Comment by erwincoumans** [comment]
I wonder why the bugs happen? Is it because some files (SimpleCamera.cpp?) appear in multiple libraries that are used in a single executable?

**Comment by RIscRIpt** [comment]
Yes, that's right, here's the full error:
```
lld-link : error : duplicate symbol: public: __cdecl SimpleCamera::SimpleCamera(void) [D:\tmp\bullet3-3.17\build\examples\TwoJoint\App_TwoJoint.vcxproj]
  >>> defined at App_TwoJoint.dir\Release\SimpleCamera.obj
  >>> defined at OpenGLWindow.lib(SimpleCamera.obj)
```
`RobotSimulator_SRCS` inside `exmaples/TwoJoint/CMakeLists.txt` contains `../../examples/OpenGLWindow/SimpleCamera.cpp`, however `App_TwoJoint` links with `OpenGLWindow` which already has this symbol defined.

Simply removing `SimpleCamera.cpp` from the list of sources doesn't seem to help, anyway there're some other places where it happens. Unfortunately CMakeLists scripts of this project seem too messy for me to try to dig into them and provide a fix with PR.

If there's an interest in full list of errors, here they are: [bullet3-lld-errors.txt](https://github.com/bulletphysics/bullet3/files/7490698/bullet3-lld-errors.txt)

**Comment by erwincoumans** [comment]
Removing SimpleCamera.cpp from the CMakeLists.txt fixed App_TwoJoint on clang-cl. Thanks for the report!
